src/pykeen/models/inductive/inductive_nodepiece_gnn.py

```diff
@@ -0,0 +1,138 @@
+        assert train_factory is not None, "train_factory must be a valid triples factory"
+        self.register_buffer(name="training_edge_index", tensor=train_factory.mapped_triples[:, [0, 2]].t())
+        self.register_buffer(name="training_edge_type", tensor=train_factory.mapped_triples[:, 1])
+
+        if inference_factory is not None:
+            inference_edge_index = inference_factory.mapped_triples[:, [0, 2]].t()
+            inference_edge_type = inference_factory.mapped_triples[:, 1]
+
+            self.register_buffer(name="validation_edge_index", tensor=inference_edge_index)
+            self.register_buffer(name="validation_edge_type", tensor=inference_edge_type)
+            self.register_buffer(name="testing_edge_index", tensor=inference_edge_index)
+            self.register_buffer(name="testing_edge_type", tensor=inference_edge_type)
+        else:
+            assert (
+                validation_factory is not None and test_factory is not None
+            ), "Validation and test factories must be triple factories"
+            self.register_buffer(name="validation_edge_index", tensor=validation_factory.mapped_triples[:, [0, 2]].t())
+            self.register_buffer(name="validation_edge_type", tensor=validation_factory.mapped_triples[:, 1])
+            self.register_buffer(name="testing_edge_index", tensor=test_factory.mapped_triples[:, [0, 2]].t())
+            self.register_buffer(name="testing_edge_type", tensor=test_factory.mapped_triples[:, 1])
+
+    def reset_parameters_(self):
+        """Reset the GNN encoder explicitly in addition to other params."""
+        super().reset_parameters_()
+        if getattr(self, "gnn_encoder", None) is not None:
+            for layer in self.gnn_encoder:
+                if hasattr(layer, "reset_parameters"):
+                    layer.reset_parameters()
+
+    def _get_representations(
+        self,
+        h: Optional[torch.LongTensor],
+        r: Optional[torch.LongTensor],
+        t: Optional[torch.LongTensor],
+        mode: InductiveMode = None,
+    ) -> Tuple[HeadRepresentation, RelationRepresentation, TailRepresentation]:
+        """Get representations for head, relation and tails, in canonical shape with a GNN encoder."""
+        entity_representations = self._entity_representation_from_mode(mode=mode)
+
+        # Extract all entity and relation representations
+        x_e, x_r = entity_representations[0](), self.relation_representations[0]()
+
+        # Perform message passing and get updated states
+        for layer in self.gnn_encoder:
+            x_e, x_r = layer(
+                x_e=x_e,
+                x_r=x_r,
+                edge_index=getattr(self, f"{mode}_edge_index"),
+                edge_type=getattr(self, f"{mode}_edge_type"),
+            )
+
+        # Use updated entity and relation states to extract requested IDs
+        # TODO I got lost in all the Representation Modules and shape casting and wrote this ;(
```

Comment:
Does this need to be addressed?

Reply:
I refactored a bit in 49a5527 and 7e40c9e.

The real solution to this would be bit complex:
* we need four `CompGCNRepresentation` instances, one for each mode.
* they all need to share the GNN modules
* they all have their own edge_index/edge_type
* they all share the same base relation representations
* the transductive training representations have their own base entity representations, and all other inductive representations share the same inductive entity representations

Reply:
fine with me to skip that if we just copy your text into the comment for future reference

Reply:
I should have been more precise - the `inference` graph is used at evaluation as well, it is a shortcut for cases when `validation` and `test` graphs are the same. Otherwise, `inference` is None, and validation/test are different graphs